hibernate-core/src/main/java/org/hibernate/metamodel/model/domain/internal/JpaMetamodelImpl.java

```diff
@@ -92,23 +102,30 @@ public TypeConfiguration getTypeConfiguration() {
 
 	@Override
 	public <X> EntityDomainType<X> resolveHqlEntityReference(String entityName) {
-		// todo (6.0) : currently we lookup the Class reference here twice potentially - fix that
-		final String rename = resolveImportedName( entityName );
-		if ( rename != null ) {
-			entityName = rename;
+		Class<X> loadedClass = null;
+		final ImportInfo<X> importInfo = resolveImport( entityName );
+		if ( importInfo != null ) {
+			loadedClass = importInfo.loadedClass;
+			entityName = importInfo.importedName;
 		}
 
 		final EntityDomainType<X> entityDescriptor = entity( entityName );
 		if ( entityDescriptor != null ) {
 			return entityDescriptor;
 		}
 
-		final Class<X> requestedClass = resolveRequestedClass( entityName );
-		if ( requestedClass != null ) {
-			return resolveEntityReference( requestedClass );
+		if ( loadedClass == null ) {
+			loadedClass = resolveRequestedClass( entityName );
+			// populate class cache for boot metamodel imports
+			if ( importInfo != null && loadedClass != null ) {
+				importInfo.loadedClass = loadedClass;
+			}
+		}
```

Comment:
The above code snippet is only meant to populate class cache for boot metamodel imports. Not sure whether it is needed. If not, the above code snippet can be deleted.